contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c

```diff
@@ -356,21 +357,34 @@ tdeheap_xlog_seg_read(int fd, void *buf, size_t count, off_t offset,
 	return readsz;
 }
 
-/* IV: TLI(uint32) + XLogRecPtr(uint64)*/
-static inline void
-SetXLogPageIVPrefix(TimeLineID tli, XLogRecPtr lsn, char *iv_prefix)
+/* IV: (TLI(uint32) + XLogRecPtr(uint64)) XOR BaseIV(uint8[12]) */
+static void
+CalcXLogPageIVPrefix(TimeLineID tli, XLogRecPtr lsn, const unsigned char *base_iv, char *iv_prefix)
 {
-	iv_prefix[0] = (tli >> 24);
-	iv_prefix[1] = ((tli >> 16) & 0xFF);
-	iv_prefix[2] = ((tli >> 8) & 0xFF);
-	iv_prefix[3] = (tli & 0xFF);
-
-	iv_prefix[4] = (lsn >> 56);
-	iv_prefix[5] = ((lsn >> 48) & 0xFF);
-	iv_prefix[6] = ((lsn >> 40) & 0xFF);
-	iv_prefix[7] = ((lsn >> 32) & 0xFF);
-	iv_prefix[8] = ((lsn >> 24) & 0xFF);
-	iv_prefix[9] = ((lsn >> 16) & 0xFF);
-	iv_prefix[10] = ((lsn >> 8) & 0xFF);
-	iv_prefix[11] = (lsn & 0xFF);
+	/* Temporary variables to make GCC vectorize it */
+	char		a[16],
+				b[16];
+
+	memset(a, 0, 16);
+	a[0] = tli >> 24;
+	a[1] = tli >> 16;
+	a[2] = tli >> 8;
+	a[3] = tli;
+
+	memset(b, 0, 16);
+	b[4] = lsn >> 56;
+	b[5] = lsn >> 48;
+	b[6] = lsn >> 40;
+	b[7] = lsn >> 32;
+	b[8] = lsn >> 24;
+	b[9] = lsn >> 16;
+	b[10] = lsn >> 8;
+	b[11] = lsn;
+
+	for (int i = 0; i < 16; i++)
```

Comment:
WAL uses CTR, the IV there should be an increasing sequence from the starting point - we should add the two numbers together.

This would work, but this with this we are even further with "standard" CTR than before. Ideally, we should do an IV calculation that would allow decryption with openssl AES-CTR algorithm. We do not verify this currently anywhere.

Reply:
Yeah, that is a good point and something I thought about. And you have conviced me now that it should be fixed. I am also veyr tempted to just hack together using CTR form OpenSSL. It should not be much work at all.